gui/gm-unit.lua

```diff
@@ -613,6 +613,700 @@ function editor_wounds:init( args )
+  
+  local startIndex = rng:random(6) -- Will give a number between 0-5 which, when accounting for the fact that the range table starts at 0, gives us the index of which of the first 6 to use
+  
+  -- Set the ranges
+  local min = selected.modifier.entry.ranges[startIndex]
+  local max = selected.modifier.entry.ranges[startIndex+1]
+  
+  -- Get the difference between the two
+  local difference = math.abs(min - max)
+  
+  -- Use the minimum, the difference, and a random roll to work out the new value.
+  local roll = rng:random(difference+1) -- difference + 1 because we want to include the max value as an option
+  local value = min + roll
+  
+  -- Set the modifier to the new value
+  if self.partChoice.type == "part" then
+    self:setPartModifier(selected.modifier.idx, value)
+  else
+    self:setBodyModifier(selected.modifier.index, value)
+  end
+end
+
+function editor_body_modifier:step(amount)
+  local _, selected = self.subviews.modifiers:getSelected()
+
+  -- Build a table of description ranges
+  local ranges = {}
+  for index, value in ipairs(selected.modifier.entry.desc_range) do
+    -- Only add a new entry if: There are none, or the value is higher than the previous range
+    if #ranges == 0 or value > ranges[#ranges] then
+      table.insert(ranges, value)
+    end
+  end
+  
+  -- Now determine what range the modifier currently falls into
+  local currentValue = selected.value
+  local rangeIndex
+  
+  for index, value in ipairs(ranges) do
+    if ranges[index+1] then -- There's still a next entry
+      if currentValue < ranges[index+1] then -- The current value is less than the next entry
+        rangeIndex = index
+        break
+      end
+    else -- This is the last entry
+      rangeIndex = index
+    end
+  end
+  
+  -- Finally, move the modifier's value up / down in range tiers based on given amount
+  local newTier = math.min(#ranges, math.max(1, rangeIndex + amount)) -- Clamp values to not go beyond bounds of ranges
+  local newValue = ranges[newTier]
+  
+  if self.partChoice.type == "part" then
+    self:setPartModifier(selected.modifier.idx, newValue)
+  else
+    self:setBodyModifier(selected.modifier.index, newValue)
+  end
+end
+
+function editor_body_modifier:updateChoices()
+  local choices = {}
+  
+  for index, modifier in ipairs(self.partChoice.modifiers) do
+    local currentValue
+    if self.partChoice.type == "part" then
+      currentValue = self.target_unit.appearance.bp_modifiers[modifier.idx[1]]
+    else -- Body
+      currentValue = self.target_unit.appearance.body_modifiers[modifier.index]
+    end
+    table.insert(choices, {text = self:beautifyString(df.appearance_modifier_type[modifier.entry.type]) .. ": " .. currentValue, value = currentValue, modifier = modifier})
+  end
+  
+  self.subviews.modifiers:setChoices(choices)
+end
+
+function editor_body_modifier:init(args)
+  self.target_unit = args.target_unit
+  self.partChoice = args.partChoice
```

Comment:
In cases like this, you can use `class.ATTRS={name=default}` with DEFAULT_NIL if you want it to be nil by default. That way if you pass something to `args` it gets auto-set to a value. More about it here: https://dfhack.readthedocs.io/en/stable/docs/Lua%20API.html#class

Reply:
Ahah, so that's how that works. I'll remember this for the future!